File: ASSI/linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
    
class ll:
    def __init__(self):
        self.head = None
    
    def add(self, data):
        n = Node(data)
        if self.head is None:
            self.head = n
        else:
            t = self.head
            while t.next is not None:
                t = t.next
            t.next = n

    def addf(self , data):
        n = Node(data)
        n.next = self.head
        self.head = n;
    
    def view(self):
        if self.head is None:
            print('empty')
        else:
            t = self.head
            while t is not None:
                print(t.data, end=' -> ')
                t = t.next
            print('None')
    
    def search(self, v):
        pos = 0
        t = self.head
        while t is not None:
            if t.data == v:
                print(f'{v} found at {pos}')
                return
            t = t.next
            pos += 1
        print('not found')
    
    def remove(self, v):
        if self.head is None:
            print('empty')
            return
        if self.head.data == v:
            self.head = self.head.next
            print('removed')
            return
        
        t = self.head
        while t.next is not None:
            if t.next.data == v:
                t.next = t.next.next
                print('removed')
                return
            t = t.next
        print('not found')

l = ll()
l.add(1)
l.add(2)
l.add(3)
l.view()         
l.search(2)      
l.remove(2)
l.view()         
l.search(2)

File: ASSI/linkedlist.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data; // . + * == ->
        this->next = nullptr; // this.next* == this->next ( same same but diff : | )
    }
};

class LinkedList {
private:
    Node* head;

public:
    LinkedList() {
        head = nullptr;
    }

    void add(int data) {
        Node* n = new Node(data);
        if (head == nullptr) {
            head = n;
        } else {
            Node* t = head;
            while (t->next != nullptr) {
                t = t->next;
            }
            t->next = n;
        }
    }
    void addf(int data){
        Node* n = new Node(data);
        n-> next = head;
        head = n;
    }
    void view() {
        if (head == nullptr) {
            cout << "empty" << endl;
        } else {
            Node* t = head;
            while (t != nullptr) {
                cout << t->data << " -> ";
                t = t->next;
            }
            cout << "None" << endl;
        }
    }

    void search(int v) {
        int pos = 0;
        Node* t = head;
        while (t != nullptr) {
            if (t->data == v) {
                cout << v << " found at " << pos << endl;
                return;
            }
            t = t->next;
            pos++;
        }
        cout << "not found" << endl;
    }

    void remove(int v) {
        if (head == nullptr) {
            cout << "empty" << endl;
            return;
        }
        if (head->data == v) {
            Node* toDelete = head;
            head = head->next;
            delete toDelete;
            cout << "removed" << endl;
            return;
        }

        Node* t = head;
        while (t->next != nullptr) {
            if (t->next->data == v) {
                Node* toDelete = t->next;
                t->next = t->next->next;
                delete toDelete;
                cout << "removed" << endl;
                return;
            }
            t = t->next;
        }
        cout << "not found" << endl;
    }
};

int main() {
    LinkedList l;
    l.addf(1);
    l.addf(2);
    l.addf(3);
    l.view();          
    l.search(2);       
    l.remove(2);
    l.view();        
    l.search(2);       

    return 0;
}